lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -2331,6 +2331,90 @@ static LogicalResult checkConnectFlow(Operation *connect) {
   return success();
 }
 
+namespace {
+enum class ConstLeafKind : uint8_t { None = 0, Normal = 1 << 0, Flip = 1 << 1 };
+} // end anonymous namespace
+
+// NOLINTBEGIN(misc-no-recursion)
+/// Checks if the type has any 'const' leaf elements that are normal or flip
+static ConstLeafKind typeConstLeafKinds(FIRRTLBaseType type,
+                                        bool outerTypeIsConst = false,
+                                        bool isFlip = false) {
+  auto typeIsConst = outerTypeIsConst || type.isConst();
+
+  if (typeIsConst && type.isPassive())
+    return isFlip ? ConstLeafKind::Flip : ConstLeafKind::Normal;
+
+  if (auto bundleType = type.dyn_cast<BundleType>()) {
+    auto kinds = ConstLeafKind::None;
+    for (auto &element : bundleType.getElements()) {
+      auto elementKinds = typeConstLeafKinds(element.type, typeIsConst,
+                                             isFlip ^ element.isFlip);
+      kinds = static_cast<ConstLeafKind>(static_cast<uint8_t>(kinds) |
+                                         static_cast<uint8_t>(elementKinds));
+    }
+    return kinds;
+  }
+
+  if (auto vectorType = type.dyn_cast<FVectorType>())
+    return typeConstLeafKinds(vectorType.getElementType(), typeIsConst, isFlip);
+
+  if (typeIsConst)
+    return isFlip ? ConstLeafKind::Normal : ConstLeafKind::Flip;
+  return ConstLeafKind::None;
+}
+// NOLINTEND(misc-no-recursion)
+
+/// Checks that connections to 'const' destinations are not dependent on
+/// non-'const' conditions in when blocks.
+static LogicalResult checkConnectConditionality(FConnectLike connect) {
+  auto dest = connect.getDest();
+  auto destType = dest.getType().dyn_cast<FIRRTLBaseType>();
+  auto src = connect.getSrc();
+  auto srcType = src.getType().dyn_cast<FIRRTLBaseType>();
+  if (!destType || !srcType)
+    return success();
+
+  auto checkConstConditionality = [&](Value value,
+                                      FIRRTLBaseType type) -> LogicalResult {
+    auto *definingBlock = value.getParentBlock();
```

Comment:
This is not the correct block, e.g.
```
module Foo:
  input p : UInt<1>
  output a : const {b : UInt<8>}
  a.b <= UInt<8>(0)
  when p:
    ; this will look valid because the defining block
    ; is the same as the connect block.
    a.b <= UInt<8>(1)
 ```
 
 You will have to walk up the subfield operations to the original declaration location.

Reply:
Ah, good catch. This also means the declaration type needs to be checked for constness as opposed to the field type since a subaccess can result in a non-const field from a const vector. This also needs to be checked regardless of when blocks to make sure non-const to const connections don't occur. 

Updated to do so with added tests.